* **Figure 2**: ASTRAL-33%-FAA tree
  * [Alignments](alignments/alignments-FAA-masked.tar.bz)
  * [Gene trees](genetrees/Best.33.FAA.tre)
  * Species tree: in [newick](speciestrees/trees/astral_trees_33_percent-FAA/estimated_species_tree.tree), in [nexus](speciestrees/SupFigs1-3-FigTree.tre) (first of the three trees)
      * For associated ASTRAL log file, see [this](speciestrees/trees/astral_trees_33_percent-FAA/astral.33.503.log).
* **Figure 3**:
  * [Scripts](https://github.com/smirarab/1kp/tree/master/scripts/hypo-test) to run the analyses are provided. 
    Moreover, the actual quarter freqencies draw on contracted newick tree files are provided. 
* **Figures S1-S3**: 
  * Nexus file: [nexus](speciestrees/SupFigs1-3-FigTree.tre)
  * Concatenation alignment used in Figure S2 is given [here](FAA-upp-masked.fasta.mask10sites.mask33taxa.tar.bz)
* **Figure S10**: Generated using DiscoVista, with inputs given [here](speciestrees/parameters/) and outputs given [here](speciestrees/output/).  
* **Figure S11**: GC contents
   - Drawn using [plot.gc.R](scripts/stats/plot.gc.R)
   - Data built using [gc-stats.py](scripts/stats/gc-stats.py)
   - Figure available [here](misc/pTpP_GC_box.pdf)
   - The name of species used in the figure are available [here](misc/names.csv).
* **Figure S12**: taxon occupancy
   - Figure drawn using [this script](scripts/stats/taxonOccupancyMap.R)
   - Data built using [this script](scripts/stats/occupancy.sh)
   - Figure available [here](misc/taxon-occupancy.png)


